### PR TITLE
ambient: rewrite controller using `krt`

### PR DESCRIPTION
**Pull request by howardjohn**
Previously this was https://github.com/istio/istio/pull/45426 but GitHub won't let me re-use it

This PR fully rewrites the ambient controller to be built on `krt`. This was one of the primary intended use cases for `krt`, so it fits very well into the existing framework.

In this PR, we move out the ambient controller to its own `ambient` package to keep it a bit more isolated, make testing simpler, and make it easier to organize things.
Most changes are isolated to the `ambient` package, though there are a few other changes:
* A variety of types are slightly tweaked, mostly to wrap with methods like ResourceName, Equals, etc. Many types are changed to not be pointers because they are just wrapping pointers anyways (so pointer is slower and more error prone)
* There are a few tests with slight tweaks. **All of these are fixing bugs**, some being that we were expecting inefficient pushes, or we forgot to wait for an event (which is probably flaky on master)
* ConfigController is no longer needed as the ambient controller creates direct kclient instances
* Add a new "RecomputeTrigger" into `krt`. This provides an escape hatch to safely rely on external state within `krt`. This would requires us to not have to rewrite then entire world with `krt` (yet?)

Within the ambient controller, this is basically a total rewrite. I wouldn't bother trying to compare diffs, as basically everything is changed.

### CODE REVIEW

**Comment by hzxuzhonghu** [comment]
Just to be clear wether a decision should be made to adopt krt widely before review. cc @costinm @ramaraochavali @linsun @keithmattix

**Comment by yuval-k** [comment]
My 2 cents here of why I'm pro this change: We recently fixed an [index bug](https://github.com/istio/istio/pull/48417) that led to policy not being enforced; With this change, it would be impossible to make these kinds of bugs again.

File: pilot/pkg/serviceregistry/kube/controller/ambient/services.go
Comment (stevenctl) [inline pilot/pkg/serviceregistry/kube/controller/ambient/services.go:42]: Is it possibly better to pass `ctx` around so the `MarkDependant` can be closer to the place the dependency actually lives?

File: pilot/pkg/serviceregistry/kube/controller/ambient/ambientindex.go
Comment (stevenctl) [inline pilot/pkg/serviceregistry/kube/controller/ambient/ambientindex.go:306]: Couldn't we make `NetworkManager` into a singleton that depends on Namespaces and MeshConfig/MeshNetworks? It's a decent amount of code, but would help hold off introducing manual recomputation triggers.

Comment (stevenctl) [reply pilot/pkg/serviceregistry/kube/controller/ambient/ambientindex.go:306]: I'd be happy to do this as a follow-up in addition to setting workloads' network gateway field

**Comment by stevenctl** [comment]
@howardjohn the passing tests inspire a decent level of confidence. Metrics/benchmarks showing this doesn't add a ton of extra computation time, unnecessary pushes, etc would be awesome. 

I'm still working my way through the newly added code but so far looks great.

File: pilot/pkg/serviceregistry/kube/controller/ambient/ambientindex.go
Comment (howardjohn) [reply pilot/pkg/serviceregistry/kube/controller/ambient/ambientindex.go:306]: Yeah I was just trying to make this PR not absolutely massive and not have 2 codepaths for network (or 1, but with stable functionality tied to krt -- right now krt only touches alpha code)

File: pilot/pkg/serviceregistry/kube/controller/ambient/workloads.go
Comment (stevenctl) [inline pilot/pkg/serviceregistry/kube/controller/ambient/workloads.go:67]: readability: can some of these fetch/filters/transformations be split into helpers?

**Comment by howardjohn** [comment]
30k pods, 2k services. krt uses a bit more memory, equal CPU. The increased memory is primarily due to List not indexing on namespace, which has a TODO already and isn't too hard to fix in a followup. This has zero proxy connections to avoid too much noise

IMO this is good enough and we can optimize as needed going forward. The cost of XDS serving is orders of magnitude higher than this especially if we have envoys involved

File: pilot/pkg/serviceregistry/kube/controller/ambient/workloads.go
Comment (bleggett) [inline pilot/pkg/serviceregistry/kube/controller/ambient/workloads.go:51]: Echoing @stevenctl a bit here - this is still effectively a series of very large, behaviorally-critical funcs that are somewhat intricate, and not unit-tested _here_ AFAICT (I am sure we are indirectly testing this with integ tests elsewhere). 

That makes this section of code a bit more intimidating to work on and validate in isolation than it needs to be.

I would like to see unit test coverage/test-based encoding of expected behavior of e.g. the Workload construction.

**Review by bleggett — APPROVED** [review]
LGTM.

I agree with most of @stevenctl 's comments about followups.

I also think `pilot/pkg/serviceregistry/kube/controller/ambient/workloads.go` needs a local `workloads_test.go` + additional decomposition, as it's still basically a couple of stupidly-large funcs which makes it difficult to grok/work on/bugfix.

This is clearly better than what we have tho, and the existing tests are all passing, so I'm good.

File: pilot/pkg/serviceregistry/kube/controller/ambient/ambientindex.go
Comment (ilrudie) [reply pilot/pkg/serviceregistry/kube/controller/ambient/ambientindex.go:91]: Don't really mind either way but I suppose `a` was chosen for ambient

File: pilot/pkg/model/service.go
Comment (howardjohn) [reply pilot/pkg/model/service.go:948]: This mirrors the logic we have in the core controller code: https://github.com/istio/istio/blob/710d15c13416700a5aacbed067f93d1defd62236/pilot/pkg/serviceregistry/kube/controller/controller.go#L722

Comment (howardjohn) [reply pilot/pkg/model/service.go:948]: You can only have overlap if one is UDP I think, and we don't support UDP

File: pilot/pkg/model/service.go
Comment (howardjohn) [reply pilot/pkg/model/service.go:1025]: metav1.LabelSelector is much more complex than LabelSelector and has the matchExpresion logic

File: pilot/pkg/serviceregistry/kube/controller/ambient/ambientindex.go
Comment (howardjohn) [reply pilot/pkg/serviceregistry/kube/controller/ambient/ambientindex.go:147]: This does a full scan though so its slow; the first check may be less common but its very cheap. This also matches the current behavior

File: pilot/pkg/serviceregistry/kube/controller/ambient/ambientindex.go
Comment (howardjohn) [reply pilot/pkg/serviceregistry/kube/controller/ambient/ambientindex.go:154]: done

File: pilot/pkg/serviceregistry/kube/controller/ambient/ambientindex.go
Comment (howardjohn) [reply pilot/pkg/serviceregistry/kube/controller/ambient/ambientindex.go:147]: actually i fixed the full scan. But I still think its best to keep the current behavior to avoid too many changes on this large pr

File: pilot/pkg/serviceregistry/kube/controller/ambient/meshconfig.go
Comment (howardjohn) [reply pilot/pkg/serviceregistry/kube/controller/ambient/meshconfig.go:41]: Fixed

File: pilot/pkg/serviceregistry/kube/controller/ambient/workloads.go
Comment (howardjohn) [reply pilot/pkg/serviceregistry/kube/controller/ambient/workloads.go:51]: We do have almost full coverage  via the ambientindex_tests.go. I can look into how we can isolate things to make them completely trivial stateless tests maybe

File: pilot/pkg/serviceregistry/kube/controller/ambient/ambientindex.go
Comment (howardjohn) [reply pilot/pkg/serviceregistry/kube/controller/ambient/ambientindex.go]: Strictly speaking for local variables there is no concept of 'exported'. But I agree it can be confusing/not standard/etc

Comment (dhawton) [reply pilot/pkg/serviceregistry/kube/controller/ambient/ambientindex.go]: Agree, just trying to think of quick glance, first time exposure, etc.

File: pilot/pkg/model/service.go
Comment (hzxuzhonghu) [reply pilot/pkg/model/service.go:1025]: fair

File: pilot/pkg/serviceregistry/kube/controller/ambient/ambientindex.go
Comment (hzxuzhonghu) [reply pilot/pkg/serviceregistry/kube/controller/ambient/ambientindex.go:280]: sg

File: pilot/pkg/serviceregistry/kube/controller/ambient/workloads.go
Comment (hzxuzhonghu) [inline pilot/pkg/serviceregistry/kube/controller/ambient/workloads.go:368]: how is the service TargetPort set, why do we find the targetPort again below?

File: pilot/pkg/serviceregistry/kube/controller/ambient/workloads.go
Comment (hzxuzhonghu) [inline pilot/pkg/serviceregistry/kube/controller/ambient/workloads.go:326]: This is not consistent with pod, which only compare with targetPortName if it is not a int type.

**Review by hzxuzhonghu — COMMENTED** [review]
The target port process does not seem right to me

File: pilot/pkg/serviceregistry/kube/controller/ambient/workloads.go
Comment (howardjohn) [reply pilot/pkg/serviceregistry/kube/controller/ambient/workloads.go:326]: :+1:

File: pilot/pkg/serviceregistry/kube/controller/ambient/workloads.go
Comment (howardjohn) [reply pilot/pkg/serviceregistry/kube/controller/ambient/workloads.go:377]: I improved the comments and added a new test. See https://github.com/istio/istio/pull/49162/commits/99ab625274525289167afd36ef5e55425161d01a. Thanks!

I think the logic here was correct just not well described; the workloadEntry part was incorrect, though.

Comment (howardjohn) [reply pilot/pkg/serviceregistry/kube/controller/ambient/workloads.go:394]: Its complex because Service and ServiceEntry behave differently

**Review by hzxuzhonghu — COMMENTED** [review]
LGTM, will approve after rebase

**Comment by howardjohn** [comment]
any chance you want to approve now to avoid time zone delays? I'll rebase tomorrow. No worries if not

**Review by hzxuzhonghu — APPROVED** [review]

**Comment by ericvn** [comment]
LGTM but waiting on approval to make sure @dhawton agrees as well.

**Review by dhawton — APPROVED** [review]
LGTM

**Pull request merged**
